### PR TITLE
APPSRE-9898 export SAPM publisher data to S3

### PR DESCRIPTION
**Pull request by fishi0x01**
Export SAPM publisher data to S3

```json
{
  "{saas-name}/{resource-template}/{target-name}/{cluster}/{namespace}":
  {
    "commit_sha": "1234",
    "deployment_state": "success|failed|missing"
  },
  ...
}
```

### CODE REVIEW

**Comment by rporres** [comment]
Who's the intended user of this data? Because it looks like we're going to have a good amount of json documents difficult to navigate in.

I may be completely off here but what I'm thinking of is: would it make sense for the destination (S3, Postgres, Mongo, whatever) to be something configurable and make the exporter something more generic that calls the different exporters based on the configuration?

**Comment by fishi0x01** [comment]
@rporres afaik we currently have a team around @jmelis that is interested in that data. cc @maorfr 

Current idea is to have a single json file containing all the information. 

I like your idea of configurable exporters - it would be a much better design. Still, due to time constraints, Im currently leaning more towards supporting that at a later point once there is a need.

Another intent of this effort is to provide an entrypoint for SAPM data export that can be easily adjusted while Im on PTO.

**Comment by rporres** [comment]
fair enough!

**Review by rporres — APPROVED** [review]

File: reconcile/saas_auto_promotions_manager/integration.py [hunk 1]
```diff
@@ -133,11 +139,16 @@ def init_external_dependencies(
     deployment_state = PromotionState(
         state=init_state(integration=OPENSHIFT_SAAS_DEPLOY, secret_reader=secret_reader)
     )
+    s3_exporter = S3Exporter(
+        state=init_state(integration=QONTRACT_INTEGRATION, secret_reader=secret_reader),
```
Comment (hemslo) [inline reconcile/saas_auto_promotions_manager/integration.py:143]: `state` requires call `state.cleanup()` after use to close aws client.

Comment (fishi0x01) [reply reconcile/saas_auto_promotions_manager/integration.py:143]: Good catch! Fixing here https://github.com/app-sre/qontract-reconcile/pull/4162

**Comment by fishi0x01** [comment]
@patjlm yes adding a "first_seen" timestamp is a good idea. My remaining question is whether that timestamp should be updated on re-deploys (e.g., test saas file with changed params) - but that can be adjusted easily as we need it.

**Pull request merged**